clients/drcachesim/tests/snoop_filter_test.cpp

```diff
@@ -0,0 +1,205 @@
+    cache_t **snooped_caches_;
+};
+
+void
+unit_test_snoop_filter_two_cores(void)
+{
+    snoop_filter_test_t snoop_filter_test(2);
+    snoop_filter_test.initialize_caches_and_snoop_filter();
+    snoop_filter_test.request_and_check_state(CPU_0, ADDR_A, TRACE_TYPE_READ,
+                                              /*num_writes*/ 0,
+                                              /*num_invalidates*/ 0,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 0,
+                                              /*block_is_dirty*/ false);
+    snoop_filter_test.request_and_check_state(CPU_1, ADDR_A, TRACE_TYPE_READ,
+                                              /*num_writes*/ 0,
+                                              /*num_invalidates*/ 0,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 1, false);
+    snoop_filter_test.request_and_check_state(CPU_0, ADDR_A, TRACE_TYPE_WRITE,
+                                              /*num_writes*/ 1,
+                                              /*num_invalidates*/ 1,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 2, false);
+    snoop_filter_test.request_and_check_state(CPU_1, ADDR_A, TRACE_TYPE_READ,
+                                              /*num_writes*/ 1,
+                                              /*num_invalidates*/ 1,
+                                              /*num_writebacks*/ 1,
+                                              /*num_sharers*/ 1, true);
+}
+
+void
+unit_test_snoop_filter_four_cores(void)
+{
+    snoop_filter_test_t snoop_filter_test(4);
+    snoop_filter_test.initialize_caches_and_snoop_filter();
+    snoop_filter_test.request_and_check_state(CPU_0, ADDR_A, TRACE_TYPE_READ,
+                                              /*num_writes*/ 0,
+                                              /*num_invalidates*/ 0,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 0,
+                                              /*block_is_dirty*/ false);
+    snoop_filter_test.request_and_check_state(CPU_1, ADDR_A, TRACE_TYPE_READ,
+                                              /*num_writes*/ 0,
+                                              /*num_invalidates*/ 0,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 1, false);
+    snoop_filter_test.request_and_check_state(CPU_2, ADDR_A, TRACE_TYPE_WRITE,
+                                              /*num_writes*/ 1,
+                                              /*num_invalidates*/ 2,
+                                              /*num_writebacks*/ 0,
+                                              /*num_sharers*/ 2, false);
```

Comment:
I again am confused by the sharers after the write, and after further reads below.